onnx/defs/data_type_utils.h

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
+// Licensed under the MIT license.
+
+#ifndef ONNX_DATA_TYPE_UTILS_H
+#define ONNX_DATA_TYPE_UTILS_H
+
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "onnx/onnx.pb.h"
+
+namespace onnx
+{
+    // String pointer as unique TypeProto identifier.
+    typedef const std::string* DTYPE;
+
+    namespace Utils
+    {
+        // Simple class which contains pointers to external string buffer and a size.
+        // This can be used to track a "valid" range/slice of the string.
+        // Caller should ensure StringRange is not used after external storage has
+        // been freed.
+        class StringRange
+        {
+        public:
+            StringRange();
+            StringRange(const char* p_data, size_t p_size);
+            StringRange(const std::string& p_str);
+            StringRange(const char* p_data);
+            const char* Data() const;
+            size_t Size() const;
+            bool Empty() const;
+            char operator[](size_t p_idx) const;
+            void Reset();
+            void Reset(const char* p_data, size_t p_size);
+            void Reset(const std::string& p_str);
+            bool StartsWith(const StringRange& p_str) const;
+            bool EndsWith(const StringRange& p_str) const;
+            bool LStrip();
+            bool LStrip(size_t p_size);
+            bool LStrip(StringRange p_str);
+            bool RStrip();
+            bool RStrip(size_t p_size);
+            bool RStrip(StringRange p_str);
+            bool LAndRStrip();
+            void ParensWhitespaceStrip();
+            size_t Find(const char p_ch) const;
+
+            // These methods provide a way to return the range of the string
+            // which was discarded by LStrip(). i.e. We capture the string
+            // range which was discarded.
+            StringRange GetCaptured();
+            void RestartCapture();
+
+        private:
+            // m_data + size tracks the "valid" range of the external string buffer.
+            const char* m_data;
+            size_t m_size;
+
+            // m_start and m_end track the captured range.
+            // m_end advances when LStrip() is called.
+            const char* m_start;
+            const char* m_end;
+        };
+
+        // Singleton wrapper around allowed data types.
+        // This implements construct on first use which is needed to ensure
+        // static objects are initialized before use. Ops registration does not work
+        // properly without this.
+        class TypesWrapper
+        {
+        public:
+
+            static TypesWrapper& GetTypesWrapper();
+
+            // DataType strings. These should match the DataTypes defined in Data.proto
```

Comment:
Totally agree, that's why I put these type strings here, as a centralized place for whole code base's reference. map you mean a "type enum" to "type string" map, right? this is not needed for now. How about we create that here later if needed please? Thank you!

Reply:
Oh, I just meant that there are 4 places in the .cc file that do remapping back and forth. So it might be easier to put them into a map once and reuse it everywhere